### ISSUE TITLE
PC-relative LDR instructions don't respect Alignment requirement

### ISSUE THREAD

**Issue by nindoja** [Vector35/arch-armv7#67]
Per the ARM spec, LDR instructions that offset from PC must 4-byte Align PC before adding the offset. 
<img width="883" alt="image (14)" src="https://user-images.githubusercontent.com/920348/177573680-f0510c50-6760-4d6b-a4f6-b2212d3fa68f.png">

It looks like, for some LDR instructions, the computed target is off by 2 bytes (sometimes +2, sometimes -2). In the screenshot below, based on manual decoding of the bytes, I think the instruction at 0x08091a1e should point to 0x8091a42 (which is a valid address pointing to the data section) but it shows up as 0x8091a44.
<img width="438" alt="image (15)" src="https://user-images.githubusercontent.com/920348/177599634-23e84e27-782b-4376-8376-5a3c59cfc839.png">

Digging into the arch plugin, I don't see the `Align` function being applied to the PC, which I think is the issue.

Minimal repro binary (manually modified a hello-world program to make this nonsense main function): 
[repro.zip](https://github.com/Vector35/arch-armv7/files/9056589/repro.zip)

For the repro above, at 0x103ca,  the LDR instruction should read from 0x103d6 (  `0x103ca (address) + 4 (PC offset) - 2 (4-byte align PC) + 10 (ldr offset)` ) but instead it reads from 0x103d8.

**Comment by lwerdna** [comment]
I think you're right. I thought non-aligned offsets in the instruction (the `#0x22` in `ldr r0, [pc, #0x22]` or the `#0xa` in `ldr r0, [pc, #0xa]`) or non-aligned effective addresses from which to load a 32-bit word were (`0x8091a42` or `0x103d6`) forbidden. But the following tools and behaviors seem to lean in favor of them being allowed.

**GNU assembler**

Trying to assemble these manually gives:

> Error: invalid offset, target not word aligned (0x00000032)

while allowing aligned offsets (0x20, 0x24, ...).

**kstool**

Kstool accepts any alignment.

**encodings**

In the 16-bit T1 encoding, imm8 has 0b00 appended, forcing alignment, but in the T2 encoding, imm12 is simply used. That they omitted the appending leads me to think they intentionally allowed unaligned offsets.

**unicorn emulator**

It loads from 0x8091a42 like you said it should.

**calculation**

Though I can only get GNU as to assemble aligned, I did many examples and the effective address was computed correctly by:

```python
def compute(pc, offset):
    return ((pc + 4) & 0xFFFFFFFC) + offset
```

And if I apply that to both your examples, it produces the offset you said is right.

Let me run a few more tests before correcting this in our source.

**Comment by lwerdna** [comment]
Here's a screenshot of the fix from the affected area in the repro binary:

![fix](https://user-images.githubusercontent.com/474191/178060500-24983284-819b-47a9-8468-d964f75faf1c.png)

Give our build machines a couple hours then check on dev update channel.

**Comment by nindoja** [comment]
Thanks @lwerdna . Will keep an eye out for a newer dev release and will let you know how it looks.

**Comment by lwerdna** [comment]
I verified the issue is resolved in 3.1.3558-dev.

Thank you for the bug report, and double thanks for linking it up with the other issue #51.

I think this has a high likelihood of not popping up again, as the fix is at an addressing mode rather than at affected instructions, where we run the risk of not having caught them all. But if you observe it again, feel free to reopen the ticket and dm me on slack.
